### PR TITLE
[core] Make LanguageRegistry not static

### LINKED ISSUE #3918: [core] Make LanguageRegistry non static
**Is your feature request related to a problem? Please describe.**
LanguageRegistry currently only exposes static methods. It would be a more useful concept if we could have different instances of LanguageRegistry, managing a different set of languages.

This is a subtask of #3782

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

- LanguageRegistry static methods are turned into instance methods
- A default instance called `LanguageRegistry.PMD` is available. With #3919 there will probably be another registry `CPD`
- The method `LanguagueRegistry::getDefaultLanguage` is removed (refs #3779)
- A new method `Language::getId` is added. I would like to replace `getTerseName` with that, to make it clear that it's an Id and not a user-facing name.

## Related issues

- Fixes #3918 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 34 violations,<br />
introduces 33 new violations, 0 new errors and 0 new configuration errors,<br />
removes 23 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/3fba1d628345496aba4e4f28ad89a4bd/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 699666 new violations, 22 new errors and 0 new configuration errors,<br />
removes 827036 violations, 39 errors and 6 configuration errors.<br />
<a href="https://chunk.io/pmd/3fba1d628345496aba4e4f28ad89a4bd/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/AbstractApexRule.java [hunk 1]
```diff
@@ -5,19 +5,13 @@
...
-import net.sourceforge.pmd.lang.LanguageRegistry;
-import net.sourceforge.pmd.lang.apex.ApexLanguageModule;
 import net.sourceforge.pmd.lang.apex.ast.ApexParserVisitor;
 import net.sourceforge.pmd.lang.ast.Node;
 import net.sourceforge.pmd.lang.rule.AbstractRule;
 
 public abstract class AbstractApexRule extends AbstractRule
         implements ApexParserVisitor {
 
-    public AbstractApexRule() {
```
Comment (oowekyala) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/AbstractApexRule.java:17]: (note: these removals are the continuation of https://github.com/pmd/pmd/pull/2617)

**Review by adangel — APPROVED** [review]
Thanks, looks good!

I'm going to merge this now... Note: This will create conflicts with #4059 

@oowekyala Do you have any idea about the changed violations in the regression report? It's only the rule LawOfDemeter...

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageVersionDiscoverer.java [hunk 2]
```diff
@@ -35,10 +36,22 @@ public LanguageVersionDiscoverer() {
...
-    public LanguageVersionDiscoverer(LanguageVersion forcedVersion) {
+    public LanguageVersionDiscoverer(LanguageRegistry registry, LanguageVersion forcedVersion) {
+        this.languageRegistry = registry;
         this.forcedVersion = forcedVersion;
     }
 
+    /**
+     * Build a new instance with no forced version.
+     *
+     * @param forcedVersion If non-null, all files should be assigned this version.
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageVersionDiscoverer.java:47]: Fixed in https://github.com/pmd/pmd/commit/2072c8e08b7d2997cfcab778cbf423f7ccf586b5

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/BaseLanguageModule.java [hunk 3]
```diff
@@ -33,10 +33,17 @@ public BaseLanguageModule(String name,
                               String terseName,
                               String firstExtension,
                               String... otherExtensions) {
+        this(name, shortName, terseName, CollectionUtil.listOf(firstExtension, otherExtensions));
+    }
+
+    public BaseLanguageModule(String name,
+                              String shortName,
+                              String terseName,
+                              List<String> extensions) {
```
[the remark this answers is not in the document]
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/BaseLanguageModule.java:42]: #4060 replaces the constructors with a new API that does this check

**Comment by oowekyala** [comment]
> @oowekyala Do you have any idea about the changed violations in the regression report? It's only the rule LawOfDemeter...

LawOfDemeter has been on my radar for a while... I think the problem is it relies accidentally on the order of rulechain visits, which is not necessarily stable across runs...

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/BaseLanguageModule.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/BaseLanguageModule.java:42]: ah, ok

**Pull request merged**